test/gtest/common/test_helpers.cc

```diff
@@ -607,4 +607,50 @@ message_stream::~message_stream() {
 
 } // detail
 
+template<typename T>
+void cartesian_product(std::vector<std::vector<T> > &final_output,
+                       std::vector<T> &cur_output,
+                       typename std::vector<std::vector<T> >
+                       ::const_iterator cur_input,
+                       typename std::vector<std::vector<T> >
+                       ::const_iterator end_input) {
+    if (cur_input == end_input) {
+        final_output.push_back(cur_output);
+        return;
+    }
+
+    const std::vector<T> &cur_vector = *cur_input;
```

Comment:
minor: maybe do "cur_input++" here, instead of '+1' in line 627?

Reply:
good point, done